rust/datafusion/src/physical_plan/repartition.rs

```diff
@@ -48,9 +51,13 @@ pub struct RepartitionExec {
     input: Arc<dyn ExecutionPlan>,
     /// Partitioning scheme to use
     partitioning: Partitioning,
-    /// Channels for sending batches from input partitions to output partitions
-    /// there is one entry in this Vec for each output partition
-    channels: Arc<Mutex<Vec<(Sender<MaybeBatch>, Receiver<MaybeBatch>)>>>,
+    /// Channels for sending batches from input partitions to output partitions.
+    /// Key is the partition number
+    channels: Arc<
```

Comment:
Note @Dandandan  asked about using a  `HashMap` vs some other structure. It is a `HashMap` for reasons explained by @edrevo here: https://github.com/edrevo/arrow/commit/97c256c4f76b8185311f36a7b27e317588904a3a#r47626396

Reply:
I didn't follow the answer completely; why do we need to remove it? 
I think `UnboundedSender` is cheap to clone and cheap to keep in memory for the duration of `RepartitionStream`?

But maybe I would need to play with the code more to see why it is needed per se.

Reply:
So ok if it's needed, but I just wasn't sure :+1:

Reply:
Thanks @Dandandan 

The use is here: https://github.com/apache/arrow/pull/9605/files#diff-b9b79e3b35bc8bfb43040ada3a4382bd0a0017ca1b1e8135be8fb310ff095674R229 

Basically this code sets up all input and outputs channels for all of the partitions and then hands out one receiver at a time in some arbitrary order (depending on the `partition` argument).

 `UnboundedReceiver` https://docs.rs/tokio/1.2.0/tokio/sync/mpsc/struct.UnboundedReceiver.html doesn't implement `Clone` (as it is multiple producer *single* consumer)

I suspect with some more thought a different structure could be used, but I couldn't convince myself it was a valuable use of time.